File: c/ccwg-api-client.c
/*
 * ccwg-api-client.c — CommonCrawl WebGraph API client (C + libcurl)
 *
 * COMPILE:
 *   cc -o ccwg-api-client ccwg-api-client.c -lcurl
 *
 * RUN:
 *   ./ccwg-api-client
 *
 * COMPILE AND RUN:
 *   cc -o ccwg-api-client ccwg-api-client.c -lcurl && ./ccwg-api-client
 *
 * REQUIRES:
 *   libcurl development headers and library. HTTPS requires a TLS library
 *   so raw sockets are not practical; libcurl is available on virtually
 *   every Linux, macOS, and FreeBSD system.
 *     macOS:   already installed (ships with the OS)
 *     Debian:  apt install libcurl4-openssl-dev
 *     RHEL:    yum install libcurl-devel
 *     FreeBSD: pkg install curl
 *
 * DESCRIPTION:
 *   Sends a POST request to the CCWG API with the hostnames listed below
 *   (one per line) and prints the raw JSON response to stdout. No JSON
 *   parsing is performed — the response is printed as-is.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <curl/curl.h>


/* --- Hostnames to look up (one per line) --- */

static const char *hostnames = "google.com\nreddit.com";


/* --- API endpoint --- */

static const char *api_url = "https://ccwg-api.customdatasets.com/";


/*--------------------------------------------------------------------------*/
static size_t write_callback(void *ptr, size_t size, size_t nmemb, void *userdata)
{
	size_t total = size * nmemb;
	(void)userdata;
	fwrite(ptr, 1, total, stdout);
	return total;
}


/*--------------------------------------------------------------------------*/
int main(void)
{
	CURL *curl;
	CURLcode res;
	long http_code = 0;

	curl_global_init(CURL_GLOBAL_DEFAULT);
	curl = curl_easy_init();

	if (!curl) {
		fprintf(stderr, "Failed to initialize curl\n");
		curl_global_cleanup();
		return 1;
	}

	curl_easy_setopt(curl, CURLOPT_URL, api_url);
	curl_easy_setopt(curl, CURLOPT_POST, 1L);
	curl_easy_setopt(curl, CURLOPT_POSTFIELDS, hostnames);
	curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, (long)strlen(hostnames));
	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
	curl_easy_setopt(curl, CURLOPT_TIMEOUT, 30L);
	curl_easy_setopt(curl, CURLOPT_CONNECTTIMEOUT, 10L);

	res = curl_easy_perform(curl);

	if (res != CURLE_OK) {
		fprintf(stderr, "\ncurl error: %s\n", curl_easy_strerror(res));
		curl_easy_cleanup(curl);
		curl_global_cleanup();
		return 1;
	}

	curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
	if (http_code != 200) {
		fprintf(stderr, "\nHTTP %ld\n", http_code);
	}

	printf("\n");

	curl_easy_cleanup(curl);
	curl_global_cleanup();
	return 0;
}

File: php/ccwg-api-client.php
<?php
declare(strict_types=1);

/*
 * ccwg-api-client.php — CommonCrawl WebGraph API client (PHP + curl)
 *
 * RUN:
 *   php ccwg-api-client.php
 *
 * REQUIRES:
 *   PHP with the curl extension (enabled by default in most distributions)
 *
 * DESCRIPTION:
 *   Sends a POST request to the CCWG API with the hostnames listed below
 *   (one per line) and prints the JSON response with pretty formatting.
 */


// --- Hostnames to look up (one per line) ---

$hostnames = "google.com\nreddit.com";


// --- API endpoint ---

$api_url = 'https://ccwg-api.customdatasets.com/';


// --- Send request ---

$ch = curl_init($api_url);
curl_setopt($ch, CURLOPT_POST, true);
curl_setopt($ch, CURLOPT_POSTFIELDS, $hostnames);
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
curl_setopt($ch, CURLOPT_TIMEOUT, 30);
curl_setopt($ch, CURLOPT_CONNECTTIMEOUT, 10);

$response = curl_exec($ch);

if ($response === false) {
	fprintf(STDERR, "curl error: %s\n", curl_error($ch));
	curl_close($ch);
	exit(1);
}

$http_code = (int)curl_getinfo($ch, CURLINFO_HTTP_CODE);
curl_close($ch);

if ($http_code !== 200) {
	fprintf(STDERR, "HTTP %d\n", $http_code);
}


// --- Pretty-print JSON response ---

$json = json_decode($response, true);

if ($json === null) {
	echo $response . "\n";
} else {
	echo json_encode($json, JSON_PRETTY_PRINT) . "\n";
}
